File: src/main/java/org/gradle/builds/model/BuildTree.java
package org.gradle.builds.model;

import java.util.List;

public interface BuildTree<T> {
    /**
     * Returns the main build for this model.
     */
    T getMainBuild();

    /**
     * Returns all of the builds that make up this tree.
     */
    List<? extends T> getBuilds();

    /**
     * Returns all of the git repos that make up this tree.
     */
    List<? extends GitRepo> getRepos();
}

File: src/main/kotlin/org/gradle/builds/model/BuildTree.kt
package org.gradle.builds.model

interface BuildTree<T> {
    /**
     * Returns the main build for this model.
     */
    val mainBuild: T

    /**
     * Returns all of the builds that make up this tree.
     */
    val builds: List<@JvmWildcard T>

    /**
     * Returns all of the git repos that make up this tree.
     */
    val repos: List<@JvmWildcard GitRepo>
}
